src/server/cluster/cluster_config_test.cc

```diff
@@ -437,29 +431,68 @@ TEST_F(ClusterConfigTest, ConfigSetMigrations) {
     }
   ])json";
 
-  auto config1 = ClusterConfig::CreateFromConfig("id0", ParseJson(config_str));
+  auto config1 = ClusterConfig::CreateFromConfig("id0", config_str);
   EXPECT_EQ(
-      config1->GetOutgoingMigrations(),
-      (std::vector<ClusterConfig::MigrationInfo>{
+      config1->GetNewOutgoingMigrations(nullptr),
+      (std::vector<MigrationInfo>{
           {.slot_ranges = {{7000, 8000}}, .target_id = "id1", .ip = "127.0.0.1", .port = 9001}}));
 
-  EXPECT_TRUE(config1->GetIncomingMigrations().empty());
+  EXPECT_TRUE(config1->GetFinishedOutgoingMigrations(nullptr).empty());
+  EXPECT_TRUE(config1->GetNewIncomingMigrations(nullptr).empty());
+  EXPECT_TRUE(config1->GetFinishedIncomingMigrations(nullptr).empty());
 
-  auto config2 = ClusterConfig::CreateFromConfig("id1", ParseJson(config_str));
+  auto config2 = ClusterConfig::CreateFromConfig("id1", config_str);
   EXPECT_EQ(
-      config2->GetIncomingMigrations(),
-      (std::vector<ClusterConfig::MigrationInfo>{
+      config2->GetNewIncomingMigrations(nullptr),
+      (std::vector<MigrationInfo>{
           {.slot_ranges = {{7000, 8000}}, .target_id = "id1", .ip = "127.0.0.1", .port = 9001}}));
 
-  EXPECT_TRUE(config2->GetOutgoingMigrations().empty());
+  EXPECT_TRUE(config2->GetFinishedOutgoingMigrations(nullptr).empty());
+  EXPECT_TRUE(config2->GetNewOutgoingMigrations(nullptr).empty());
+  EXPECT_TRUE(config2->GetFinishedIncomingMigrations(nullptr).empty());
+
+  auto config3 = ClusterConfig::CreateFromConfig("id2", config_str);
+  EXPECT_TRUE(config3->GetFinishedOutgoingMigrations(nullptr).empty());
+  EXPECT_TRUE(config3->GetNewIncomingMigrations(nullptr).empty());
+  EXPECT_TRUE(config3->GetFinishedIncomingMigrations(nullptr).empty());
+  EXPECT_TRUE(config3->GetNewOutgoingMigrations(nullptr).empty());
+
+  const auto* config_str2 = R"json(
+  [
+    {
+      "slot_ranges": [ { "start": 0, "end": 6999 } ],
+      "master": { "id": "id0", "ip": "localhost", "port": 3000 },
+      "replicas": []
+    },
+    {
+      "slot_ranges": [ { "start": 7000, "end": 16383 } ],
+      "master": { "id": "id1", "ip": "localhost", "port": 3001 },
+      "replicas": []
+    }
+  ])json";
+
+  auto config4 = ClusterConfig::CreateFromConfig("id0", config_str2);
+  auto config5 = ClusterConfig::CreateFromConfig("id1", config_str2);
 
-  auto config3 = ClusterConfig::CreateFromConfig("id2", ParseJson(config_str));
-  EXPECT_TRUE(config3->GetIncomingMigrations().empty());
-  EXPECT_TRUE(config3->GetOutgoingMigrations().empty());
+  EXPECT_EQ(
+      config4->GetFinishedOutgoingMigrations(config1),
+      (std::vector<MigrationInfo>{
+          {.slot_ranges = {{7000, 8000}}, .target_id = "id1", .ip = "127.0.0.1", .port = 9001}}));
```

Comment:
Nit: If you add a gtest comparer for `MigrationInfo` you'll be able to use something like `EXPECT_THAT(..., ElementsAre(...))` which will have better outputs in failure cases, and will also make the comparison statement easier to read

Reply:
Config is really simple, if I add some new functionality I will add gtest comparer